sematic/resolvers/tests/test_local_resolver.py

```diff
@@ -308,7 +310,40 @@ def pipeline():
     runs_by_id = {run.id: run for run in runs}
 
     assert runs_by_id[future.id].future_state == FutureState.NESTED_FAILED.value
-    assert runs_by_id[future.id].exception == "Failed because the child run failed"
+    assert runs_by_id[future.id].exception == ExceptionMetadata(
+        repr="Failed because the child run failed",
+        name="Exception",
+        module="builtins",
+    )
 
     assert runs_by_id[future.nested_future.id].future_state == FutureState.FAILED.value
-    assert "FAIL!" in runs_by_id[future.nested_future.id].exception
+    assert "FAIL!" in runs_by_id[future.nested_future.id].exception.repr
+
+
+_tried = 0
+
+
+class SomeException(Exception):
+    pass
+
+
+@func(retry=RetrySettings(exceptions=(SomeException,), times=3))
+def try_three_times():
+    global _tried
+    _tried += 1
+    raise SomeException()
+
+
+@mock_no_auth
+def test_retry(test_db, mock_requests, valid_client_version):  # noqa: F811
+    future = try_three_times()
+    try:
+        future.resolve(LocalResolver())
+    except SomeException:
+        pass
+    else:
+        assert False
+
+    assert future.props.retry_settings.retry_count == 3
+    assert future.state == FutureState.FAILED
+    assert _tried == 4
```

Comment:
Shouldn't `tries` be like a counting number rather than an index? I.e. 1 try=> no retries if failure, just attempt once. 2 tries => execute, and if it fails do it exactly once more. In that case I would expect `_tried` to be `3` here after calling `try_three_times`